core/src/test/java/apoc/convert/ConvertJsonTest.java

```diff
@@ -782,14 +839,57 @@ public void testToTreeLeafNodesWithConfigErrorInclude() {
         }
     }
 
+    @Test
+    public void testToTreeDoesNotRemoveNonDuplicateRels() {
+        String createStatement = """
+            CREATE (v1:N {id: 'n21', name: 'Node 21', p2: 'node21'})
+            CREATE (v2:N {id: 'n22', name: 'Node 22', p2: 'node22'})
+            CREATE (v1)-[:R {prop1: 'n21->n22 [1]', prop2: 'Rel 1'}]->(v2)
+            CREATE (v1)-[:R {prop1: 'n21->n22 [2]', prop2: 'Rel 2'}]->(v2)""";
+
+        db.executeTransactionally(createStatement);
+
+        String query = """
+                MATCH p1 = (n:N {id:'n21'})-[e1]->(m1:N)
+                WITH  COLLECT(p1) as paths
+                CALL apoc.convert.toTree(paths, false)
+                YIELD value
+                RETURN value""";
+
+        testResult(
+                db,
+                query,
+                res -> {
+                    Map<String, Object> value = (Map<String, Object>) res.next().get("value");
+                    // Check Parent Node: (v1:T1 {id: 'v21', name: 'Vertex 21', p2: 'value21'})
+                    assertEquals("n21", value.get("id"));
+                    assertEquals("N", value.get("_type"));
+                    assertEquals("node21", value.get("p2"));
+                    assertEquals("Node 21", value.get("name"));
+                    // Check Children Nodes
+                    List<Map<String, Object>> RBranches = (List<Map<String, Object>>) value.get("R");
+                    Map<String, Object> rel1 = Map.of("p2", "node22", "name", "Node 22", "id", "n22", "R.prop1", "n21->n22 [1]", "R.prop2", "Rel 1");
+                    Map<String, Object> rel2 = Map.of("p2", "node22", "name", "Node 22", "id", "n22", "R.prop1", "n21->n22 [2]", "R.prop2", "Rel 2");
+                    assertEquals(2, RBranches.size());
+                    if (RBranches.get(0).entrySet().containsAll(rel1.entrySet())) {
+                        assertTrue(RBranches.get(1).entrySet().containsAll(rel2.entrySet()));
+                    } else {
+                        assertTrue(RBranches.get(0).entrySet().containsAll(rel1.entrySet()));
+                        assertTrue(RBranches.get(1).entrySet().containsAll(rel2.entrySet()));
+                    }
+                    assertFalse(res.hasNext());
```

Comment:
I have recently learned to appreciate `org.assertj.core.api.Assertions.assertThat` assertions (heavily used in the java parts of the mono repo). Their API is very vast, and it's easy to get lost, https://assertj.github.io/doc/#assertj-overview gives a quick overview what it is about.

The main benefit, for me, is improved error messages, if you don't want to use the full extent of the API it is often useful at least wrap the existing assertions. In this example that could look like for example:

```java
org.assertj.core.api.Assertions.assertThat(res.stream().toList()).satisfiesExactly(row -> {
    Map<String, Object> value = (Map<String, Object>) row.get("value");
    // The existing assertions, except assertFalse(res.hasNext()); (not needed because satisfiesExactly)
});
```

If this fails it will not only tell you `x was not equal to y`, but give you the complete context `Expected all entries in <the complete result> to satisfy all requirements but failed with: x was not equal to y`. Super helpful when tests start failing because you get the complete picture.

Anyway, I don't expect you to change(!), but if that sounds nice, you could consider to use assertj at some other time :).

Reply:
Ah this was interesting, I will keep in mind for the next APOC task! :D